cmd/cerbosctl/hub/store/upload_git.go

```diff
@@ -0,0 +1,314 @@
+	"github.com/go-git/go-git/v5/plumbing/object"
+
+	"github.com/cerbos/cerbos/internal/util"
+)
+
+const uploadGitHelp = `
+The following exit codes have a special meaning.
+	- 6: The version condition supplied using --version-must-eq wasn't satisfied
+
+# Apply the file changes recorded in the git repo between commit 55a4248 and HEAD
+
+cerbosctl hub store upload-git 55a4248
+cerbosctl hub store upload-git 55a4248 --path path/to/git/repository
+
+# Apply the file changes recorded in the git repo between commit 55a4248 to e746228
+
+cerbosctl hub store upload-git 55a4248 e746228
+cerbosctl hub store upload-git 55a4248 e746228 --path path/to/git/repository
+`
+
+type UploadGitCmd struct {
+	diffToApply   *diff
+	From          string `arg:"" help:"Git revision to start from when generating the diff"`
+	To            string `arg:"" help:"Git revision to end when generating the diff (The resolved reference must be the ancestor of the from argument)" default:"HEAD"`
+	Path          string `help:"Path to the git repository" default:"."`
+	Subdirectory  string `help:"Subdirectory under the given path to check and upload changes from" aliases:"subdir" default:"."`
+	Output        `embed:""`
+	ChangeDetails `embed:""`
+	VersionMustEq int64 `help:"Require that the store is at this version before committing the change" optional:""`
+}
+
+func (*UploadGitCmd) Help() string {
+	return uploadGitHelp
+}
+
+func (ugc *UploadGitCmd) Validate() error {
+	r, err := git.PlainOpen(ugc.Path)
+	if err != nil {
+		return fmt.Errorf("failed to open git repository: %w", err)
+	}
+
+	from, err := r.ResolveRevision(plumbing.Revision(ugc.From))
+	if err != nil {
+		return fmt.Errorf("failed to resolve revision %q: %w", ugc.From, err)
+	}
+
+	to, err := r.ResolveRevision(plumbing.Revision(ugc.To))
+	if err != nil {
+		return fmt.Errorf("failed to resolve revision %q: %w", ugc.To, err)
+	}
+
+	if ugc.diffToApply, err = ugc.diff(r, *from, *to); err != nil {
+		return fmt.Errorf("failed to get diff between revision %q to %q: %w", ugc.From, ugc.To, err)
+	}
+
+	return nil
+}
+
+func (ugc *UploadGitCmd) Run(k *kong.Kong, cmd *Cmd) error {
+	repository, err := git.PlainOpen(ugc.Path)
+	if err != nil {
+		return fmt.Errorf("failed to open git repository: %w", err)
+	}
+
+	client, err := cmd.storeClient()
+	if err != nil {
+		return ugc.toCommandError(k.Stderr, err)
+	}
+
+	version := ugc.VersionMustEq
+	for batch, err := range ugc.batch() {
+		if err != nil {
+			return ugc.toCommandError(k.Stderr, err)
+		}
+
+		gitChangeDetails, err := changeDetailsFromHash(repository, ugc.diffToApply.hash)
+		if err != nil {
+			return ugc.toCommandError(k.Stderr, fmt.Errorf("failed to get change details for %q: %w", ugc.diffToApply.hash.String(), err))
+		}
+
+		changeDetails, message, err := ugc.ChangeDetails.ChangeDetails(gitChangeDetails)
+		if err != nil {
+			return ugc.toCommandError(k.Stderr, fmt.Errorf("failed to get change details: %w", err))
+		}
```

Comment:
These can be moved outside the loop because they'd generated the same value every time.